### PR TITLE
add support for Intel USM extension to chipStar OpenCL BE

### PR DESCRIPTION
**Pull request by franz**
disabled by default, can be enabled with cmake option `-DINTEL_USM_SUPPORT=1`

### CODE REVIEW

**Comment by pvelesko** [comment]
conflicts + please let me know when ready for review

**Comment by franz** [comment]
rebased & reformatted, should now be ready for review

File: src/backend/OpenCL/CHIPBackendOpenCL.hh [hunk 1]
```diff
@@ -119,20 +119,32 @@ public:
   cl::Program *get();
 };
 
+typedef struct {
+  clSharedMemAllocINTEL_fn clSharedMemAllocINTEL;
+  clDeviceMemAllocINTEL_fn clDeviceMemAllocINTEL;
+  clHostMemAllocINTEL_fn clHostMemAllocINTEL;
+  clMemFreeINTEL_fn clMemFreeINTEL;
+} CHIPContextUSMExts;
```
Comment (linehill) [inline src/backend/OpenCL/CHIPBackendOpenCL.hh:127]:
```suggestion
struct CHIPContextUSMExts {
  clSharedMemAllocINTEL_fn clSharedMemAllocINTEL;
  clDeviceMemAllocINTEL_fn clDeviceMemAllocINTEL;
  clHostMemAllocINTEL_fn clHostMemAllocINTEL;
  clMemFreeINTEL_fn clMemFreeINTEL;
};
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -189,7 +189,7 @@ option(CHIP_USE_EXTERNAL_HIP_TESTS "Use Catch2 tests from the hip-tests submodul
 option(CHIP_USE_OCML_ROUNDED_OPS "Use OCML implementations for devicelib functions with explicit rounding mode such as __dadd_rd. Otherwise, rounding mode will be ignored" OFF)
 option(CHIP_ENABLE_NON_COMPLIANT_DEVICELIB_CODE "Enable non-compliant devicelib code such as calling LLVM builtins from inside kernel code. Enables certain unsigned long devicelib func variants" OFF)
 option(CHIP_FAST_MATH "Use native_ OpenCL functions which are fast but their precision is implementation defined" OFF)
-
+option(INTEL_USM_SUPPORT "enable support for cl_intel_unified_shared_memory in the OpenCL backend" OFF)
```
Comment (pjaaskel) [inline CMakeLists.txt:192]: All the other options a prefixed with CHIP_ perhaps not stop that tradition?

File: src/backend/OpenCL/CHIPBackendOpenCL.hh [hunk 3]
```diff
@@ -151,11 +163,16 @@ public:
 };
 
 class CHIPContextOpenCL : public chipstar::Context {
-public:
-  bool allDevicesSupportFineGrainSVM();
+private:
+  cl::Context ClContext;
+  bool SupportsIntelUSM;
```
Comment (pjaaskel) [inline src/backend/OpenCL/CHIPBackendOpenCL.hh:168]: Is 'Supports' a wrong word here if you want to store which API the allocation was done with; IntelUSMAllocated or similar might be more descriptive?

File: src/backend/OpenCL/SVMemoryRegion.cc [hunk 4]
```diff
@@ -24,33 +24,70 @@
...
       logTrace("clSVMFree on: {}\n", PtrToFree);
-      clSVMFree(Ctx(), PtrToFree);
+      if (SupportsUSM)
+        clMemFreeINTEL(Ctx(), PtrToFree);
+      else
+        clSVMFree(Ctx(), PtrToFree);
     };
     auto SPtr = std::shared_ptr<void>(Ptr, Deleter);
+    logTrace("Memory allocated: {} / {}\n", Ptr, Size);
     SvmAllocations_.emplace(SPtr, Size);
```
Comment (pjaaskel) [inline src/backend/OpenCL/SVMemoryRegion.cc:87]: Why emplace? Should we error out if there's an existing allocation instead?

**Review by pjaaskel — APPROVED** [review]

File: src/backend/OpenCL/SVMemoryRegion.cc (hunk 4 above)
Comment (linehill) [reply src/backend/OpenCL/SVMemoryRegion.cc:87]: SvmAllocation_ shouldn't feature the pointer we are adding unless we forget to remove pointers when they are freed. We could enforce this with an assertion.

File: src/backend/OpenCL/CHIPBackendOpenCL.hh (hunk 3 above)
Comment (franz) [reply src/backend/OpenCL/CHIPBackendOpenCL.hh:168]: fixed

File: src/backend/OpenCL/SVMemoryRegion.cc (hunk 4 above)
Comment (franz) [reply src/backend/OpenCL/SVMemoryRegion.cc:88]: added assertion

File: CHIPSPVConfig.hh.in [hunk 5]
```diff
@@ -42,4 +42,6 @@
 // Non-zero integer if the chipStar is built in debug mode.
 #cmakedefine CHIP_DEBUG_BUILD @CHIP_DEBUG_BUILD@
 
+#cmakedefine CHIP_USE_INTEL_USM
+
```
Comment (pvelesko) [inline CHIPSPVConfig.hh.in:46]: This doesn't require `@CHIP_USE_INTEL_USM` like in the line above?

Comment (franz) [reply CHIPSPVConfig.hh.in:46]: It's not necessary; value is not required for boolean CMake variables. `#cmakedefine X` turns into `#define X` if `X=ON`, and `#undef X` with `X=OFF`.

**Pull request merged**